### ISSUE TITLE
Empty slice fails to infer type when it's able to

### ISSUE THREAD

**Issue by simvux** [rust-lang/rust#67790]
Let's say we have a trait and implement that for a slice (or tuple containing slice)
```rust
trait This {
    fn that(&self) -> u8;
}

impl This for &[i32] {
    fn that(&self) -> u8 {
        0
    }
}
```
If we then attempt to use the trait with an empty slice it's unable to infer a type for the slice, causing a misleading error saying
`the trait This is not implemented for [_; 0]`
```rust
fn main() {
    let x = This::that(&[]);
}
```
However if we simply create an annotated const or let binding
```rust
fn main() {
    const EMPTY: &[i32] = &[];
    let x = This::that(&EMPTY);
}
```
It works just fine

Shouldn't rustc be able to infer the type of the slice here?

**Comment by ExpHP** [comment]
Two problems:

* I think you either meant to impl the trait for `[i32]`, or you meant to call `This::that(&&[])`.
  (I'll assume you meant to impl the trait for `[i32]`)
* Even with that fix, this still won't work.

Unification is always performed *before* coercions are inserted, meaning that coercions never occur in a generic context.  (the compiler must be able to *see* the mismatch in types without needing to look at trait impls)

The first example is seen as:

```rust
fn main() {
    let x = This::that(&[]);
}

// equivalent to
fn main() {
    // [...] literals are always array types, not slices
    let array: [?0; 0] = [];

    // Each type parameter in a generic function call gets a fresh
    // type inference variable, including the Self type. (?1 below)
    let arg: &?1 = &array;
    let func = <?1 as This>::that;
    let x = func(arg);
}
```

Type inference first encounters `let arg ...` and **unifies `&?1` with `&[?0; 0]`** to determine that `?1 = [?0; 0]`.  Then later, when checking trait bounds, it tries to find an impl associated with `<[?0; 0] as This>`, but there is none because the trait is not implemented on arrays.

When you annotate the type as a slice, you force a coercion:

```rust
fn main() {
    let annotated: &[_] = &[];
    let x = This::that(annotated);
}

// equivalent to
fn main() {
    let array: [?0; 0] = [];
    let annotated: &[?2] = &array;

    let arg: &?1 = annotated;
    let func = <?1 as This>::that;
    let x = func(arg);
}
```

At `let annotated`, the compiler tries to unify `&[?2]` and `&[?0; 0]` and fails, **so it inserts an unsizing coercion** on the rhs to turn the array into a slice (`&[?0; 0] -> &[?0]`).  This allows unification to succeed and determine that `?2 = ?0`.  Therefore, this time at the `let arg` line, **it unifies &`?1` with `&[?0]`** and determines that `?1 = [?0]`.  When it does trait checking, it finds a unique impl that unifies with this (`impl This for &[i32]`) and can thus infer that `?0 = i32`.

**Comment by ExpHP** [comment]
A more common workaround is to add `[..]` after the array:

```rust
fn main() {
    let x = This::that(&[][..]);
}
```

**Comment by jonas-schievink** [comment]
I think this is expected, since coercions won't be attempted to find an impl. In many cases, when there's only one impl, it will be selected automatically, but that does not happen when the `Self` type is mismatched.
